core/src/main/java/org/infinispan/remoting/inboundhandler/BaseBlockingRunnable.java

```diff
@@ -28,10 +28,18 @@ protected BaseBlockingRunnable(BasePerCacheInboundInvocationHandler handler, Cac
 
    @Override
    public void run() {
+      InvocationStatus invocationStatus = null;
       try {
-         response = beforeInvoke();
-         if (response == null) {
-            response = handler.invokePerform(command);
+         invocationStatus = beforeInvoke();
+         switch (invocationStatus) {
+            case CONTINUE:
+               response = handler.invokePerform(command);
+               break;
+            case CACHE_NOT_FOUND:
+               response = CacheNotFoundResponse.INSTANCE;
+               break;
+            case DEFERRED:
+               return;
```

Comment:
@danberindei Aha, I meant rather per-sender + per-key (than just per-key) ordering, but not blocking writes to keyA because update for keyB is delayed. But understood.

If releasing the lock immediately and using ordered messages is more performant, you could already adopt that for `DIST_SYNC` and use the RPC ack only to notify the originator about completion. But that could still have technical edge cases, and `DIST_ASYNC` is already expected to be less reliable.
I personally think that too much contention should not happen in practice, so optimizing that is low-priority. On the other hand, with per-sender ordering I am worried about global hiccups when a message is lost/dropped.